firmware/library/L2_HAL/sensors/temperature/temperature.hpp

```diff
@@ -1,11 +1,74 @@
-// Interface used to sense the temperature of the temperature sensor.
 #pragma once
-#include <cstdint>
+
+#include "L1_Drivers/i2c.hpp"
+#include "utility/bit.hpp"
+
 class TemperatureInterface
 {
  public:
-  virtual bool Initialize() = 0;
-  virtual uint32_t GetTemperatureBytes() = 0;
-  virtual float GetCelsius() = 0;
-  virtual float GetFahrenheit() = 0;
+  virtual bool Initialize()                        = 0;
+  virtual float GetCelsius(I2c i2c_temperature)    = 0;
+  virtual float GetFahrenheit()                    = 0;
 };
+
+class Temperature : public TemperatureInterface
+{
+ public:
+  bool Initialize() override
+  {
+  i2c_.Initialize();
+  uint8_t temperature_sensor_id = 0x00;
+  // Checks to see if the value read by I2c
+  // for the Temperature Chip ID register
+  // is the same as the Base ID (0x14).
+  i2c_.WriteThenRead(kTemperatureAddress, &kTemperatureId,
+  1, &temperature_sensor_id, 1);
+  return (temperature_sensor_id == kMemoryMapAddress);
+  }
+
+  float GetCelsius(I2c i2c_temperature) override
+  {
+  uint8_t one_burst_register         = 0xC4;
+  uint8_t autoinc_bit_register       = 0xC5;
+  uint8_t most_significant_register  = 0xC1;
+  uint8_t least_significant_register = 0xC2;
+  // The register will enable the device to collect data once
+  // and automatically sets the stop bit to 0 (2nd bit).
+  status_ = i2c_temperature.Write(kTemperatureAddress,
+ { one_burst_register, 0x04 }, 2);
+  // Auto increments I2c register address pointer.
+  status_ = i2c_temperature.Write(kTemperatureAddress,
+ { autoinc_bit_register, 0x01 }, 2);
+  i2c_temperature.WriteThenRead(kTemperatureAddress,
+ &most_significant_register, 1, &most_significant_register, 1);
+  i2c_temperature.Read(kTemperatureAddress, &least_significant_register, 1);
+
+  int bit_data = 0;
+  bit_data = ((most_significant_register & 0x7F) << 8);
+  bit_data = bit::Insert(bit_data, least_significant_register, 0, 128);
+  // The required computation after bit shifting.
+  // Formula can be found below, see page 4:
+  // datasheets/Temperature-Sensor/si7060-datasheets.pdf
+  // Note that: 1 << 14 = 2^14 = 16384
+  celsius_data_ = ((bit_data - (1 << 14)) / 160) + 55;
+  return (celsius_data_);
+  }
+
+  float GetFahrenheit() override
+  {
+  // Formula used to convert Celsius to Fahrenheit
+  // Note that: 1 << 5 = 2^5 = 32
+  fahrenheit_data_ = ((celsius_data_ - (1 << 5)) * 5) / 9;
+  return (fahrenheit_data_);
+  }
+
+ private:
+  static constexpr uint8_t kTemperatureAddress = 0x31;
+  static constexpr uint8_t kTemperatureId      = 0xC0;
+  static constexpr uint8_t kMemoryMapAddress   = 0x14;
+  float celsius_data_;
+  float fahrenheit_data_;
+  I2c i2c_;
+  Status status_;
```

Comment:
You never give the user the option to retrieve the status information nor do you actually use it in your code. You assign it, but never read it. So keeping this within the class doesn't do anything but increase the size of the object. Bare in mind that ever single byte on the MCU counts. We cannot afford unnecessary adding weight to each object when we only have 96kB of ram total.